README.md
# Electric Trains (LTN Branch)

This version of Electric Trains adds support for LTN, 
this does unfortunately need a [patched LTN](https://github.com/z0w13/Logistic-Train-Network) as well


changelog.txt
---------------------------------------------------------------------------------------------------
Version: 1.0.4
Date: 29. 06. 2020
  Features:
    - 0.18 release.
  Balancing:
    - Upgraded capacity of locomotives and batteries.
    - Increased charging rate of train stops and battery charging stations.
---------------------------------------------------------------------------------------------------
Version: 1.0.3
Date: 08. 08. 2018
  Features:
    - German translation added, courtesy of animewolf0258.
  Balancing:
    - Increased the stack size of discharged battery packs.
    - Doubled the charge rate of electric train stops and battery charging stations.
---------------------------------------------------------------------------------------------------
Version: 1.0.2
Date: 15. 03. 2018
  Bugfixes:
    - Electric train stops will now give back an electric train stop item when mined.
---------------------------------------------------------------------------------------------------
Version: 1.0.1
Date: 06. 03. 2018
  Features:
  Balancing:
    - Reduced the drain on battery charging stations.
  Bugfixes:
    - Re-enabled some code I forgot to in the control script, the EEIs associated with electric train stops will now be deleted properly upon deconstruction of the train stop.



control.lua
-- For debugging purposes
--
function table.val_to_str ( v )
  if "string" == type( v ) then
    v = string.gsub( v, "\n", "\\n" )
    if string.match( string.gsub(v,"[^'\"]",""), '^"+$' ) then
      return "'" .. v .. "'"
    end
    return '"' .. string.gsub(v,'"', '\\"' ) .. '"'
  else
    return "table" == type( v ) and table.tostring( v ) or
      tostring( v )
  end
end

function table.key_to_str ( k )
  if "string" == type( k ) and string.match( k, "^[_%a][_%a%d]*$" ) then
    return k
  else
    return "[" .. table.val_to_str( k ) .. "]"
  end
end

function table.tostring( tbl )
  local result, done = {}, {}
  for k, v in ipairs( tbl ) do
    table.insert( result, table.val_to_str( v ) )
    done[ k ] = true
  end
  for k, v in pairs( tbl ) do
    if not done[ k ] then
      table.insert( result,
        table.key_to_str( k ) .. "=" .. table.val_to_str( v ) )
    end
  end
  return "{" .. table.concat( result, "," ) .. "}"
end

function debugLog(message)
  game.print(message)
end
--]]

--[[ global table data structure
global.
  stations
  ... -- unit # as key
    {station entity, charging station entity}
  
  chargingTrains
  ... trainID as key
    {train entity, charging station entity, {list of locomotive entities}, currently charging locomotive}
--]]

remote.add_interface("electrictrains", {
  status = function()
    debugLog("Global station list: " .. table.tostring(global.stations))
    debugLog("Charging trains: " .. table.tostring(global.chargingTrains))
  end
})

local waitStation = defines.train_state.wait_station
local maxFuelValue = 250000000 -- Maximum fuel value of the electric locomotive when not consuming battery power.
local maxChargingRate = 25000000 * 23 / 60 -- Power transferred per second (energy is transferred every 23 ticks) by the control script.

script.on_init(function()
  global.stations = {}
  global.chargingTrains = {}
end)

script.on_load(function()
  if next(global.chargingTrains) ~= nil then
    script.on_nth_tick(23, tick)
  end
end)

function is_electric_strain_stop(entity)
    return entity.name == "electric-train-stop" or entity.name == "electric-logistic-train-stop"
end

function entity_built(event)
  if is_electric_strain_stop(event.created_entity) then
    local position = event.created_entity.position
    local force = event.created_entity.force
    local surface = event.created_entity.surface.index
    
    local charger = game.surfaces[surface].create_entity{name = "charging-station", position = position, force = force}
    global.stations[event.created_entity.unit_number] = {event.created_entity, charger}
  elseif event.created_entity.name == "electric-locomotive" then
    event.created_entity.burner.currently_burning = "electric-locomotive-fuel-dummy"
    event.created_entity.burner.remaining_burning_fuel = 100000
  end
end

script.on_event(defines.events.on_built_entity, entity_built)
script.on_event(defines.events.on_robot_built_entity, entity_built)

function entity_removed(event)
  if is_electric_strain_stop(event.entity)  then
    global.stations[event.entity.unit_number][2].destroy()
    global.stations[event.entity.unit_number] = nil
    
    for i,v in pairs(global.chargingTrains) do
      if not v[2].valid then
        global.chargingTrains[i] = nil
      end
    end
  end
end

script.on_event(defines.events.on_player_mined_entity, entity_removed)
script.on_event(defines.events.on_robot_mined_entity, entity_removed)
script.on_event(defines.events.on_entity_died, entity_removed)

function train_changed_state(event)
  if event.train.state == waitStation and is_electric_strain_stop(event.train.station)  then
    local locomotives = find_electric_locomotives(event.train)
    
    if #locomotives ~= nil then
      set_train_charging(event.train, event.train.station.unit_number, locomotives)
    end
  end
end

script.on_event(defines.events.on_train_changed_state, train_changed_state)

function find_electric_locomotives(luaTrain)
  local electricLocomotives = {}
  
  for i,v in pairs(luaTrain.locomotives) do
    for j,u in pairs(v) do
      if u.name == "electric-locomotive" then
        table.insert(electricLocomotives, u)
      end
    end
  end
  
  return electricLocomotives
end

function set_train_charging(trainEntity, stationIndex, locomotives)
  global.chargingTrains[trainEntity.id] = {trainEntity, global.stations[stationIndex][2], locomotives}
  activateTicker()
end

function activateTicker()
  if next(global.chargingTrains) ~= nil then
    script.on_nth_tick(23, tick)
  end
end

function tick()
  local currentFuelValue
  local chargeTransferValue
  
  for i,v in pairs(global.chargingTrains) do
    if not v[1].valid or v[1].state ~= waitStation or next(v[3]) == nil then
      global.chargingTrains[i] = nil
    else
      if v[4] == nil then
        v[4] = v[3][1]
        if v[4].burner.currently_burning ~= nil and v[4].burner.currently_burning.name ~= "electric-locomotive-fuel-dummy" then
          local burnt_result = game.item_prototypes[v[4].burner.currently_burning.name].burnt_result
          if burnt_result ~= nil then
            v[4].burner.burnt_result_inventory.insert({name = burnt_result.name})
          end
        end
        v[4].burner.currently_burning = "electric-locomotive-fuel-dummy" -- Set the 'currently burning' item to scale the locomotive's remaining fuel bar properly.
      end
      
      currentFuelValue = v[4].burner.remaining_burning_fuel
      
      if v[2] ~= nil and v[2].valid and currentFuelValue < maxFuelValue then
        chargeTransferValue = math.min(maxFuelValue - currentFuelValue, v[2].energy, maxChargingRate)
        v[2].energy = v[2].energy - chargeTransferValue
        v[4].burner.remaining_burning_fuel = currentFuelValue + chargeTransferValue
      else
        v[4] = nil
        table.remove(v[3],1)
      end
    end
  end
  
  if next(global.chargingTrains) == nil then
    script.on_nth_tick(nil)
  end
end


prototypes/items.lua
data:extend(
{
  {
    type = "item-with-entity-data",
    name = "electric-locomotive",
    icon = "__ElectricTrains__/graphics/icons/electric-locomotive.png",
    icon_size = 32,
    subgroup = "transport",
    order = "a[train-system]-g[electric-locomotive]",
    place_result = "electric-locomotive",
    stack_size = 5
  },
  
  {
    type = "item",
    name = "electric-train-stop",
    icon = "__ElectricTrains__/graphics/icons/train-stop.png",
    icon_size = 64, icon_mipmaps = 4,
    subgroup = "transport",
    order = "a[train-system]-c[train-stop]-a",
    place_result = "electric-train-stop",
    stack_size = 10
  },
  {
    type = "item",
    name = "battery-charging-station",
    icon = "__base__/graphics/icons/accumulator.png",
    icon_size = 64, icon_mipmaps = 4,
    subgroup = "production-machine",
    order = "h[train-system]",
    place_result = "battery-charging-station",
    stack_size = 10
  },
  
  {
    type = "item",
    name = "battery-pack",
    icon = "__base__/graphics/technology/battery.png",
    icon_size = 128,
    fuel_category = "electrical",
    fuel_value = "100MJ",
    burnt_result = "discharged-battery-pack",
    subgroup = "intermediate-product",
    order = "s-a[battery-pack]",
    stack_size = 10
  },
  
  {
    type = "item",
    name = "discharged-battery-pack",
    icon = "__ElectricTrains__/graphics/icons/discharged-battery.png",
    icon_size = 128,
    subgroup = "intermediate-product",
    order = "s-b[discharged-battery-pack]",
    stack_size = 10
  },
  
  {
    type = "item",
    name = "electric-locomotive-fuel-dummy",
    icon = "__base__/graphics/icons/accumulator.png",
    icons = {
      {icon = "__ElectricTrains__/graphics/icons/electric-locomotive.png", icon_size = 32},
      {icon = "__ElectricTrains__/graphics/icons/lightning-bolt.png", icon_size = 32, scale = 0.5, shift = {12, 12}}
    },
    fuel_category = "electrical",
    fuel_value = "250MJ",
    subgroup = "raw-material",
    order = "h[battery]",
    stack_size = 1
  },
  
  { -- defined to stop Factorio complaining about not having an item to place for the proper charging station
    type = "item",
    name = "charging-station",
    icon = "__base__/graphics/icons/train-stop.png",
    icon_size = 32,
    subgroup = "transport",
    order = "a[train-system]-c[train-stop]-a",
    place_result = "charging-station",
    stack_size = 10
  }
})

if data.raw["train-stop"]["logistic-train-stop"] ~= nil then
  data:extend({
    {
      type = "item",
      name = "electric-logistic-train-stop",
      icon = "__ElectricTrains__/graphics/icons/train-stop.png",
      icon_size = 64, icon_mipmaps = 4,
      subgroup = "transport",
      order = "a[train-system]-c[logistic-train-stop]-a",
      place_result = "electric-logistic-train-stop",
      stack_size = 10
    }
  })
end

prototypes/recipes.lua
data:extend(
{
  {
    type = "recipe",
    name = "electric-locomotive",
    energy_required = 5,
    enabled = false,
    ingredients =
    {
      {"electric-engine-unit", 20},
      {"advanced-circuit", 10},
      {"steel-plate", 30},
      {"battery", 50}
    },
    result = "electric-locomotive"
  },
  
  {
    type = "recipe",
    name = "electric-train-stop",
    energy_required = 5,
    enabled = false,
    ingredients =
    {
      {"steel-plate", 30},
      {"battery", 10},
      {"copper-cable", 20}
    },
    result = "electric-train-stop"
  },
  {
    type = "recipe",
    name = "battery-charging-station",
    energy_required = 5,
    enabled = false,
    ingredients =
    {
      {"iron-plate", 15},
      {"advanced-circuit", 5},
      {"copper-cable", 20}
    },
    result = "battery-charging-station"
  },
  
  {
    type = "recipe",
    name = "battery-pack",
    energy_required = 10,
    enabled = false,
    ingredients =
    {
      {"steel-plate", 2},
      {"battery", 20}
    },
    result = "battery-pack"
  },
  
  {
    type = "recipe",
    name = "battery-pack-recharge",
    category = "electrical",
    energy_required = 10,
    enabled = false,
    ingredients =
    {
      {"discharged-battery-pack", 1}
    },
    result = "battery-pack"
  }
})

if data.raw["train-stop"]["logistic-train-stop"] ~= nil then
  data:extend({
    {
      type = "recipe",
      name = "electric-logistic-train-stop",
      energy_required = 5,
      enabled = false,
      ingredients =
      {
        {"electric-train-stop", 1},
        {"constant-combinator", 1},
        {"small-lamp", 1},
        {"green-wire", 2},
        {"red-wire", 2},
      },
      result = "electric-logistic-train-stop"
    }
  })
end

prototypes/technologies.lua
data:extend(
{
  {
    type = "technology",
    name = "electric-railway",
    icon_size = 128,
    icon = "__base__/graphics/technology/railway.png",
    effects =
    {
      {
        type = "unlock-recipe",
        recipe = "electric-locomotive"
      },
      {
        type = "unlock-recipe",
        recipe = "electric-train-stop"
      },
      {
        type = "unlock-recipe",
        recipe = "electric-logistic-train-stop"
      },
      {
        type = "unlock-recipe",
        recipe = "battery-charging-station"
      },
      {
        type = "unlock-recipe",
        recipe = "battery-pack"
      },
      {
        type = "unlock-recipe",
        recipe = "battery-pack-recharge"
      }
    },
    prerequisites = {"railway", "electric-energy-distribution-2", "electric-engine"},
    unit =
    {
      count = 150,
      ingredients =
      {
        {"automation-science-pack", 1},
        {"logistic-science-pack", 1},
        {"chemical-science-pack", 1}
      },
      time = 30
    },
    order = "c-g-aa",
  }
})


prototypes/entities.lua
local locomotive = util.table.deepcopy(data.raw.locomotive["locomotive"])
locomotive.name = "electric-locomotive"
locomotive.minable.result = "electric-locomotive"
locomotive.weight = 1200
locomotive.max_speed = 1.5
locomotive.max_power = "1MW"
locomotive.reversing_power_modifier = 1.0
locomotive.burner.fuel_category = "electrical"
locomotive.burner.fuel_inventory_size = 1
locomotive.burner.burnt_inventory_size = 1
locomotive.burner.smoke = nil
locomotive.color = {r = 0, g = 0.33, b = 0.92, a = 0.5}
locomotive.stop_trigger = 
{
  {
    type = "play-sound",
    sound =
    {
      {
        filename = "__base__/sound/train-breaks.ogg",
        volume = 0.3
      }
    }
  }
}

local batteryChargingStation = util.table.deepcopy(data.raw["accumulator"]["accumulator"])
batteryChargingStation.type = "assembling-machine"
batteryChargingStation.name = "battery-charging-station"
batteryChargingStation.minable.result = "battery-charging-station"
batteryChargingStation.energy_source = {
  type = "electric",
  buffer_capacity = "20MJ",
  usage_priority = "primary-input",
  input_flow_limit = "10MW",
  output_flow_limit = "0kW",
  drain = "500W"
}
batteryChargingStation.animation = accumulator_picture()
batteryChargingStation.working_visualisations = accumulator_charge()
batteryChargingStation.working_visualisations.light = batteryChargingStation.charge_light
batteryChargingStation.crafting_categories = {"electrical"}
batteryChargingStation.crafting_speed = 1
batteryChargingStation.energy_source = {
  type = "electric",
  usage_priority = "secondary-input",
  drain = "3kW"
}
batteryChargingStation.next_upgrade = nil
batteryChargingStation.energy_usage = "10MW"
batteryChargingStation.fixed_recipe = "battery-pack-recharge"
batteryChargingStation.show_recipe_icon = false

function setup_electric_stop(entity)
  data:extend(
  {
    locomotive,
    entity,
    
    {
      type = "electric-energy-interface",
      name = "charging-station",
      icons = {
        {icon = "__ElectricTrains__/graphics/icons/train-stop.png"},
        {icon = "__ElectricTrains__/graphics/icons/lightning-bolt.png", icon_size = 32, scale = 0.5, shift = {-12, 12}}
      },
      icon_size = 64, icon_mipmaps = 4,
      flags = {"placeable-neutral", "player-creation", "not-blueprintable"},
      corpse = "small-remnants",
      -- selection_box = {{-0.5, -0.5}, {0.5, 0.5}},
      energy_source =
      {
        type = "electric",
        buffer_capacity = "25MJ",
        usage_priority = "primary-input",
        input_flow_limit = "25MW",
        output_flow_limit = "0kW",
        drain = "500W"
      },
      charge_cooldown = 45,
      discharge_cooldown = 30,
    },
    
    batteryChargingStation
  })
end


local electricTrainStop = util.table.deepcopy(data.raw["train-stop"]["train-stop"])
electricTrainStop.name = "electric-train-stop"
electricTrainStop.color = {r = 0, g = 0.33, b = 0.92, a = 0.5}
electricTrainStop.minable.result = "electric-train-stop"
setup_electric_stop(electricTrainStop)

if data.raw["train-stop"]["logistic-train-stop"] ~= nil then
  local electricLogisticTrainStop = util.table.deepcopy(data.raw["train-stop"]["logistic-train-stop"])
  electricLogisticTrainStop.name = "electric-logistic-train-stop"
  electricLogisticTrainStop.color = {r = 0, g = 0.33, b = 0.92, a = 0.5}
  electricLogisticTrainStop.minable.result = "electric-logistic-train-stop"
  setup_electric_stop(electricLogisticTrainStop)
end